#4413 HTML Report wrong steps/substeps nesting

reinholdfuereder opened the pull request:
Previously using/calling the same meta step more than once (e.g. calling the same PageObject method twice), led to rendering all the steps from the meta step nested below the first meta step execution/node again and again, instead of several meta step executions/nodes (with each having just all steps from the meta step once).

reinholdfuereder committed:
#4413 HTML Report wrong steps/substeps nesting

Previously using/calling the same meta step more than once (e.g. calling the same PageObject method twice), led to rendering all the steps from the meta step nested below the first meta step execution/node again and again, instead of several meta step executions/nodes (with each having just all steps from the meta step once).

reinholdfuereder commented:
@Naktibalda or @DavertMik How are the chances that this is merged and part of the next Codeception PHAR file release?

(See https://github.com/Codeception/Codeception/issues/4413 for details.)

Naktibalda commented:
@DavertMik Please review this pull request.
If you approve it, I will merge it to other branches and release them.

Naktibalda commented:
@reinholdfuereder It would be great if you could create a test which demonstrates this issue in Codeception repo.
Reports are tested by https://github.com/Codeception/Codeception/blob/3.0/tests/cli/RunCest.php
You can create a new test suite which demonstrates wrong step issue in https://github.com/Codeception/Codeception/tree/3.0/tests/data/claypit/tests .

I am going to release all branches soon to support PHPUnit 8.2 https://github.com/Codeception/Codeception/issues/5541
and I would like to include this change to release, but I need either tests or approval by @DavertMik

Naktibalda commented:
It was easier to fix PHPUnit 8.2 compatibility in Codeception/Codeception repo, so there is no time presure for this change again.

reinholdfuereder commented:
@Naktibalda First of all sorry for the massive delay. At long last I started with a test as you suggested above:
* maybe you can have a look at it via issue #4413 or https://github.com/reinholdfuereder/Codeception/tree/3.1 and give some initial feedback
* frankly I am not sure how to continue with the fix (which is in this PR) that is needed in "codeception/phpunit-wrapper", while the test is in "codeception/codeception": I mean how to deal with composer/versioning/etc when it is two different github projects/repositories...

(Sorry again in advance, but I won't be able to continue before Tuesday next week -- the earliest...)

Naktibalda commented:
I pushed your test to https://github.com/Codeception/Codeception/tree/issue-4413 branch.

You can make Travis run tests against that branch by setting `CODECEPTION_VERSION: 'dev-issue-4413'` in https://github.com/Codeception/phpunit-wrapper/blob/8.0/.travis.yml#L4

reinholdfuereder commented:
@Naktibalda Thanks for the help/hint => unfortunately the test still fails even with the fix (cf. PR #66):
* https://travis-ci.org/Codeception/phpunit-wrapper/jobs/586122912 vs. https://travis-ci.org/Codeception/phpunit-wrapper/jobs/586129034 (note that there are other test failures too; those must be unrelated to my changes though)
* I had a closer look at the changes, but only with respect to first different case ("Case7"): at a first glance it looks OK/good, because the new HTML report actually now contains "Page\DemoPageObject: demo action1" a second time
* unfortunately it is a bit cumbersome to try to find the problem (is the test, i.e. presumably the built regex, maybe not correct, or the fix itself?) via many commits/pushes/whatever via CI/Travis: is there an (easy) way for me to test this locally too? Because I think I would also need to play around with the test case. Am I actually allowed to push changes to this dev branch https://github.com/Codeception/Codeception/tree/issue-4413?

reinholdfuereder commented:
@Naktibalda I have found and fixed the problem in https://github.com/reinholdfuereder/Codeception/commit/3033f80434c5ebd35498cd78e4137fbc95c99932 (update: at long last I have moved this into a dedicated branch: https://github.com/reinholdfuereder/Codeception/tree/3.1-4413) -- I had a copy&paste failure in my test => I am a bit embarrassed but also surprised why...

How to proceed?
1. How can I get the updated test into https://github.com/Codeception/Codeception/tree/issue-4413?
2. How can I then re-trigger CI for https://github.com/Codeception/phpunit-wrapper/pull/66?
3. Assuming that the test then succeeds: how to proceed so that fix will make it into next Codeception 3.1.x (PHAR file!) release?
  * There is no PR for new test and test suite in Codeception/Codeception yet, but I think I could do that based on https://github.com/reinholdfuereder/Codeception/tree/3.1-4413 (it is merge-able now)
    * => OK, I lost patience or hope it is easier for you and thus have now created the PR: https://github.com/Codeception/Codeception/pull/5690
  * There are 2 PRs for Codeception/phpunit-wrapper, the first one is just the fix, while the second one was based on your hint as approach for required Codeception version dependency change for CI => so maybe the first one should be taken, while the second one should be closed; unless the test PR is first merged (and expected to fail due to the missing fix)
    * For Codeception 3.1.x PHAR file release I guess you need to merge the Codeception/phpunit-wrapper PR to potentially different branches; and/or adapt version dependencies in Codeception/Codeception!?

Thanks for your help in advance.

Naktibalda commented:
I will review this PR and respond to other questions later.

> @Naktibalda (Unrelated to this issue, but...) Because you just mentioned Codeception 4 to be released soon: Would it please be possible to merge my recent little PRs (Codeception/phpunit-wrapper#62 + #5690; #5688) for 2 IMHO rather bad Codeception bugs before and soon release them in a (last?) 3.1.x version; and of course also merge them into new Codeception 4 codebase? (This would be very motivating and also promising with respect to the future usage of Codeception...) Thanks!

Changes made in phpunit-wrapper will affect all versions of Codeception starting with, probably, 2.4 and ending with 4.0
I will take care of merging these changes to other branches.

There will be at least one more release of 3.1.

reinholdfuereder commented:
Thanks @Naktibalda!

Naktibalda commented:
I merged #66 to 8.0
